### PR TITLE
[WebFlux] Look up user data from userinfo endpoint

### LINKED ISSUE #13461: When using OIDC with WebFlux, look up the user's information from the /userinfo endpoint
This issue is similar to #11767, but for WebFlux.

##### **Overview of the feature request**

When using OAuth 2.0 / OIDC for authentication with WebFlux, the [`AccountResource#getAccount()` calls `UserService#getUserFromAuthentication()`](https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/main/java/package/web/rest/AccountResource.java.ejs#L75). 

This logic casts the principal to an `AbstractAuthenticationToken` because that's the parent for both OAuth login and a resource server. Ideally, the access token could be used from this token and the `OidcUser` can be fetched, which will contain a fully-hydrated user object.

##### **Motivation for or Use Case**

When using Ionic or React Native, JHipster is used as a resource server. Whatever is contained in the access token will be read on the server. For Ionic, you have to add several new claims for things to work nicely.

From the [Ionic for JHipster docs](https://github.com/oktadeveloper/generator-jhipster-ionic#add-claims-to-access-token):

> In order to authentication successfully with your Ionic app, you have to do a bit more configuration in Okta. Since the Ionic client will only send an access token to JHipster, you need to 1) add a `groups` claim to the access token and 2) add a couple more claims so the user's name will be available in JHipster.
> 
> Navigate to **API** > **Authorization Servers**, click the **Authorization Servers** tab and edit the **default** one. Click the **Claims** tab and **Add Claim**. Name it "groups" and include it in the Access Token. Set the value type to "Groups" and set the filter to be a Regex of `.*`. Click **Create**.
> 
> Add another claim, name it `given_name`, include it in the access token, use `Expression` in the value type, and set the value to `user.firstName`. Optionally, include it in the `profile` scope. Perform the same actions to create a `family_name` claim and use the expression `user.lastName`.

With this change, the last paragraph could be deleted.

##### **Related issues or PR**

#11767 and PR #12609.

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by mraible**
Fixes #13461. 

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary https://github.com/jhipster/jhipster.github.io/pull/1215
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
This PR is based on an Ionic + WebFlux monolith. You can see the changes from a generated JHipster app in https://github.com/mraible/ionic-jhipster/pull/3. Thanks to @rwinch's [example](https://github.com/rwinch/jhipster-13461), which made this all possible.

I still need to test this in a gateway. 

This feature still needs user caching to avoid lookups for every request. Spring MVC has a [cache implementation](https://github.com/jhipster/generator-jhipster/blob/main/generators/server/templates/src/main/java/package/security/oauth2/CustomClaimConverter.java.ejs#L73) so the lookup only happens once per user. 

I asked Rob about our Spring MVC implementation and here's what he said:

> re: once per user
> 
> I'd imagine caching in a hashmap could be bad since there is no expiry. Changing attributes would never get updated and it can grow indefinitely (run out of memory). If you are ok w/ that limitation, you can do the same on the reactive side.
> 
> btw if you want more flexible cache consider using spring cache abstraction

Regarding the `.block()` in the `jwtDecoder` bean, Rob said:

> The bean creation process is not reactive, so you have to block

/cc @Falydoor

**Comment by mraible** [comment]
@rwinch I tried adding `@Cacheable` to the `enrich()` method and making it public. 

```java
@Cacheable
public Mono<Jwt> enrich(String token, Jwt jwt) { ... }
```

That doesn't seem to work, most likely because data in the ID token JWT changes each time. It'd be slick if I could somehow make the cache key the access token. Is that possible or do I need to create a wrapper object that has the token and JWT?

**Comment by rwinch** [comment]
Sorry I'm not sure I understand what your goal is. Do you want the cache key to be the token parameter? If so, you should be able to add the [`key` property](https://docs.spring.io/spring-framework/docs/current/reference/html/integration.html#cache-annotations-cacheable-key) to customize the cache key. Something like `@Cacheable(key="#token")`

**Comment by mraible** [comment]
@rwinch I'll try that. Thanks!

**Comment by mraible** [comment]
@rwinch I tried adding `@EnableCaching` to the `SecurityConfiguration` and adding `@Cacheable` to the `enrich()` method:

```java
@Cacheable(cacheNames="users", key="#token")
public Mono<Jwt> enrich(String token, Jwt jwt) { } 
```

It still seems like this method is hit every time, even though the `token` value is the same.

**Comment by mraible** [comment]
I tried to reproduce the failure in `ms-mf-vue-consul-oauth2-mysql-memcached` and `npm run e2e` passes when I run everything with the default `dev` profile. However, if I try to start each app with the `prod` profile, the `blog` app fails to start because MySQL is already listening on 3306 for the gateway.

```
 ⠋ Container docker-blog-mysql-1  Starting                                                                         0.1s
Error response from daemon: driver failed programming external connectivity on endpoint docker-blog-mysql-1 (fe57a869c40190bbe1b514a6bd75059e593bdb10b09014f6ba8dc75560773cd2): Bind for 127.0.0.1:3306 failed: port is already allocated
```

**Comment by mraible** [comment]
@DanielFran Any idea why tests are failing?

**Comment by mshima** [comment]
It's possible to see the stack trace:

```
2022-06-07 21:20:32.590ERROR1---[or-http-epoll-1]a.w.r.e.AbstractErrorWebExceptionHandler: [8c4a550f-295]  500 Server Error for HTTP GET "/remoteEntry.js"
[93](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:95)
  
[94](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:96)
  org.springframework.web.reactive.function.client.WebClientResponseException$Unauthorized: 401 Unauthorized from GET http://keycloak:9080/realms/jhipster/protocol/openid-connect/userinfo
[95](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:97)
  	at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:198)
[96](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:98)
  	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
[97](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:99)
  Error has been observed at the following site(s):
[98](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:100)
  	*__checkpoint ⇢ 401 from GET http://keycloak:9080/realms/jhipster/protocol/openid-connect/userinfo [DefaultWebClient]
[99](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:101)
  	*__checkpoint ⇢ org.springframework.security.web.server.authentication.AuthenticationWebFilter [DefaultWebFilterChain]
[100](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:102)
  	*__checkpoint ⇢ com.okta.developer.notification.web.filter.SpaWebFilter [DefaultWebFilterChain]
[101](https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true#step:27:103)
```

Go to the run logs https://github.com/jhipster/generator-jhipster/runs/6782407626?check_suite_focus=true
Go to `Dump docker logs` section.
Click at `notification (docker-compose_notification_1)` image.

**Comment by mraible** [comment]
Thanks, @mshima! This is very helpful.

**Comment by mraible** [comment]
@mshima My first suspicion is that the token is empty so that's why calling the `/userinfo` endpoint is failing. However, if I try to print it out, the result is:

```
sub: 4c973896-5761-41fc-8217-07c5d13a004b
token: ***
```

Do you know if GitHub Actions is preventing the printing of the token value, or this is the actual value?

**Comment by mshima** [comment]
> @mshima My first suspicion is that the token is empty so that's why calling the `/userinfo` endpoint is failing. However, if I try to print it out, the [result](https://github.com/jhipster/generator-jhipster/runs/6785183437?check_suite_focus=true#step:27:299) is:
> 
> ```
> sub: 4c973896-5761-41fc-8217-07c5d13a004b
> token: ***
> ```
> 
> Do you know if GitHub Actions is preventing the printing of the token value, or this is the actual value?

Don't know. I may see if I can help latter today or tomorrow.

**Comment by mraible** [comment]
The Couchbase issues might be caused by my M1. It looks like there's a newer version of `couchbase/server` [available](https://forums.couchbase.com/t/apple-silicon-darwin-arm64-apple-m1-computer-hardware-w-mac-os-11/29189/17). I tried changing `gateway/src/main/docker/couchbase/Couchbase.Dockerfile` to  the following, but it doesn't seem to help:

```
FROM couchbase/server:7.1.0
```

If I change it to `FROM couchbase/server:7.1.0-aarch64`, it works. However, I don't think it should make the CI tests fail since they're not running on M1. I'll try changing the version to 7.1.0 and see if that helps.

File: generators/generator-constants.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ const DOCKER_POSTGRESQL_VERSION = '14.3';
 const DOCKER_POSTGRESQL = `postgres:${DOCKER_POSTGRESQL_VERSION}`;
 const DOCKER_MONGODB_VERSION = '4.4.14';
 const DOCKER_MONGODB = `mongo:${DOCKER_MONGODB_VERSION}`;
-const DOCKER_COUCHBASE_VERSION = '7.0.3';
+const DOCKER_COUCHBASE_VERSION = '7.1.0';
```
Comment (mraible) [inline generators/generator-constants.js:70]:
```suggestion
const DOCKER_COUCHBASE_VERSION = '7.0.4';
```

File: generators/generator-constants.js [hunk 2]
```diff
@@ -67,7 +67,7 @@ const DOCKER_POSTGRESQL_VERSION = '14.3';
 const DOCKER_POSTGRESQL = `postgres:${DOCKER_POSTGRESQL_VERSION}`;
 const DOCKER_MONGODB_VERSION = '4.4.14';
 const DOCKER_MONGODB = `mongo:${DOCKER_MONGODB_VERSION}`;
-const DOCKER_COUCHBASE_VERSION = '7.0.3';
+const DOCKER_COUCHBASE_VERSION = '7.0.4';
```
Comment (mraible) [inline generators/generator-constants.js:70]:
```suggestion
const DOCKER_COUCHBASE_VERSION = '7.0.3';
```

**Comment by mraible** [comment]
@mshima The fact that [this commit](https://github.com/jhipster/generator-jhipster/pull/18241/commits/10d4873ad5571d8067d7b140c34c110638109b63) fixes the Couchbase tests is a little concerning. Do some CI tests override the `authenticationType`?

**Comment by mshima** [comment]
@mraible jwtSecretKey is used for passing authentication between gateway and microservices, am I wrong?

**Comment by mraible** [comment]
`jwtSecretKey` is only used when you choose `jwt` for authentication. `session` and `oauth2` don't use it.

**Comment by mraible** [comment]
It seems strange that a comment change triggered a test failure...

**Comment by mraible** [comment]
I've started implementing a test similar to [`CustomClaimConverterIT.java.ejs`](https://github.com/jhipster/generator-jhipster/blob/main/generators/server/templates/src/test/java/package/security/oauth2/CustomClaimConverterIT.java.ejs). It'd be cool if I could test it locally, but when I try `mvn verify`, I get an error:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:testCompile (default-testCompile) on project blog: Compilation failure: Compilation failure:
[ERROR] /Users/mraible/Downloads/webflux/src/test/java/com/jhipster/demo/blog/config/MysqlTestContainer.java:[7,37] cannot find symbol
[ERROR]   symbol:   class MySQLContainer
[ERROR]   location: package org.testcontainers.containers
[ERROR] /Users/mraible/Downloads/webflux/src/test/java/com/jhipster/demo/blog/config/MysqlTestContainer.java:[16,13] cannot find symbol
[ERROR]   symbol:   class MySQLContainer
[ERROR]   location: class com.jhipster.demo.blog.config.MysqlTestContainer
[ERROR] -> [Help 1]
```

If I run `mvn verify -Dtest=AccountResourceIT -Pprod`, it fails too. This is not a test I'm working on, so I figured it'd work. Instead, it errors:

```
Factory method 'connectionFactory' threw exception; nested exception is 
org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryOptionsInitializer$ConnectionFactoryBeanCreationException: 
Failed to determine a suitable R2DBC Connection URL
```

##### **JHipster Version(s)**

```
blog@0.0.1-SNAPSHOT /Users/mraible/Downloads/webflux
└── generator-jhipster@7.8.1

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationIndex": 0,
    "applicationType": "monolith",
    "authenticationType": "oauth2",
    "baseName": "blog",
    "blueprints": [],
    "buildTool": "maven",
    "cacheProvider": "no",
    "clientFramework": "angularX",
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1654796357425,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "devServerPort": 4200,
    "dtoSuffix": "DTO",
    "enableGradleEnterprise": false,
    "enableHibernateCache": false,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "entities": ["Blog", "Post", "Tag"],
    "entitySuffix": "",
    "gradleEnterpriseHost": "",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "7.8.1",
    "languages": ["en"],
    "lastLiquibaseTimestamp": 1654796537000,
    "messageBroker": false,
    "nativeLanguage": "en",
    "otherModules": [],
    "packageFolder": "com/jhipster/demo/blog",
    "packageName": "com.jhipster.demo.blog",
    "pages": [],
    "prodDatabaseType": "mysql",
    "reactive": true,
    "searchEngine": false,
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "skipUserManagement": true,
    "testFrameworks": ["cypress"],
    "websocket": false,
    "withAdminUi": true
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}
entity Post {
  title String required
  content TextBlob required
  date Instant required
}
entity Tag {
  name String required minlength(2)
}
relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}
relationship ManyToMany {
  Post{tag(name)} to Tag{entry}
}

paginate Post, Tag with infinite-scroll

</pre>
</details>


##### **Environment and Tools**

openjdk version "17.0.3" 2022-04-19
OpenJDK Runtime Environment Temurin-17.0.3+7 (build 17.0.3+7)
OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (build 17.0.3+7, mixed mode)

git version 2.32.1 (Apple Git-133)

node: v16.15.0

npm: 8.6.0

Docker version 20.10.14, build a224086

Docker Compose version v2.5.1

**Comment by mraible** [comment]
I think it's best to 🚢 without a test to verify everything works. I spent several hours trying to implement a test today and I feel like I lost hours of my life.

**Comment by mraible** [comment]
Here's what I tried:

`SecurityConfiguration.java`:

```java
    @Bean
    ReactiveJwtDecoder jwtDecoder(ReactiveClientRegistrationRepository registrations) {
        Mono<ClientRegistration> clientRegistration = registrations.findByRegistrationId("oidc");
        WebClient webClient = WebClient.create();
        return new CustomClaimConverter(clientRegistration, webClient,
            jHipsterProperties.getSecurity().getOauth2().getAudience()).jwtDecoder();
    }
```

`CustomClaimConverter.java`:

```java
package tech.jhipster.sample.security.oauth2;

import org.springframework.core.ParameterizedTypeReference;
import org.springframework.security.oauth2.client.registration.ClientRegistration;
import org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator;
import org.springframework.security.oauth2.core.OAuth2TokenValidator;
import org.springframework.security.oauth2.jwt.*;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class CustomClaimConverter {

    private final Mono<ClientRegistration> registration;
    private final WebClient webClient;

    private final List<String> allowedAudience;

    // todo: optimize for scale https://github.com/jhipster/generator-jhipster/issues/18868
    private final Map<String, Mono<Jwt>> users = new ConcurrentHashMap<>();

    public CustomClaimConverter(Mono<ClientRegistration> registration, WebClient webClient, List<String> allowedAudience) {
        this.registration = registration;
        this.webClient = webClient;
        this.allowedAudience = allowedAudience;
    }

    public ReactiveJwtDecoder jwtDecoder() {
        return registration
            .map(oidc -> createJwtDecoder(oidc.getProviderDetails()))
            .block();
    }

    private ReactiveJwtDecoder createJwtDecoder(ClientRegistration.ProviderDetails details) {
        NimbusReactiveJwtDecoder jwtDecoder = new NimbusReactiveJwtDecoder(details.getJwkSetUri());
        OAuth2TokenValidator<Jwt> audienceValidator = new AudienceValidator(allowedAudience);
        OAuth2TokenValidator<Jwt> withIssuer = JwtValidators.createDefaultWithIssuer(details.getUserInfoEndpoint().getUri());
        OAuth2TokenValidator<Jwt> withAudience = new DelegatingOAuth2TokenValidator<>(withIssuer, audienceValidator);

        jwtDecoder.setJwtValidator(withAudience);

        return new ReactiveJwtDecoder() {
            @Override
            public Mono<Jwt> decode(String token) throws JwtException {
                return jwtDecoder.decode(token).flatMap(jwt -> enrich(token, jwt));
            }

            private Mono<Jwt> enrich(String token, Jwt jwt) {
                // Only look up user information if identity claims are missing
                if (jwt.hasClaim("given_name") && jwt.hasClaim("family_name")) {
                    return Mono.just(jwt);
                }
                // Retrieve user info from OAuth provider if not already loaded
                return users.computeIfAbsent(
                    jwt.getSubject(),
                    s -> {
                        return webClient
                            .get()
                            .uri(details.getUserInfoEndpoint().getUri())
                            .headers(headers -> headers.setBearerAuth(token))
                            .retrieve()
                            .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {
                            })
                            .map(userInfo ->
                                Jwt
                                    .withTokenValue(jwt.getTokenValue())
                                    .subject(jwt.getSubject())
                                    .audience(jwt.getAudience())
                                    .headers(headers -> headers.putAll(jwt.getHeaders()))
                                    .claims(claims -> {
                                        String username = userInfo.get("preferred_username").toString();
                                        // special handling for Auth0
                                        if (userInfo.get("sub").toString().contains("|") && username.contains("@")) {
                                            userInfo.put("email", username);
                                        }
                                        // Allow full name in a name claim - happens with Auth0
                                        if (userInfo.get("name") != null) {
                                            String[] name = userInfo.get("name").toString().split("\\s+");
                                            if (name.length > 0) {
                                                userInfo.put("given_name", name[0]);
                                                userInfo.put("family_name", String.join(" ", Arrays.copyOfRange(name, 1, name.length)));
                                            }
                                        }
                                        claims.putAll(userInfo);
                                    })
                                    .claims(claims -> claims.putAll(jwt.getClaims()))
                                    .build()
                            );
                    }
                );
            }
        };
    }
}
```

`CustomClaimConverterIT.java`:

```java
package tech.jhipster.sample.security.oauth2;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.ReactiveJwtDecoder;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;
import tech.jhipster.sample.IntegrationTest;

import java.util.HashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * This test ensures that a JWT passed in without identity claims
 * will be populated with them from the /userinfo endpoint
 */
@IntegrationTest
@ExtendWith(MockitoExtension.class)
class CustomClaimConverterIT {

    private Mono<Map<String, Object>> claims;
    private static final String USERNAME = "admin";
    private static final String NAME = "John";
    private static final String FAMILY_NAME = "Doe";
    private static final String FULL_NAME = NAME + " " + FAMILY_NAME;
    private static final String NAME_SUFFIX = "Sr.";
    private static final String EMAIL = "john.doe@gmail.com";

    private WebClient webClient = mock(WebClient.class);
    private WebClient.RequestHeadersSpec requestHeadersSpec = mock(WebClient.RequestHeadersSpec.class);
    private WebClient.RequestHeadersUriSpec requestHeadersUriSpec = mock(WebClient.RequestHeadersUriSpec.class);
    private WebClient.ResponseSpec responseSpec = mock(WebClient.ResponseSpec.class);

    @Autowired
    ReactiveClientRegistrationRepository clientRegistrationRepository;

    @Autowired
    ReactiveJwtDecoder jwtDecoder;

    private void mockHttpGetUserInfo(Map<String, Object> userInfo) {
        when(webClient.get()).thenReturn(requestHeadersUriSpec);
        when(requestHeadersUriSpec.uri("https://api.jhipster.org/user")).thenReturn(requestHeadersSpec);
        when(requestHeadersSpec.retrieve()).thenReturn(responseSpec);
        when(responseSpec.bodyToMono(Map.class)).thenReturn(Mono.just(userInfo));
    }

    @Test
    void testConvert() {
        // GIVEN
        Map<String, Object> claims = new HashMap<>();
        claims.put("sub", "123");
        // AND
        Map<String, Object> user = new HashMap<>();
        user.put("preferred_username", USERNAME);
        user.put("given_name", NAME);
        user.put("family_name", FAMILY_NAME);
        mockHttpGetUserInfo(user);

        // WHEN
        String jwtWithSub = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjMiLCJpYXQiOjE1MTYyMzkwMjJ9.n8ngm2beoF993ylp-trkEa6kV4Mgx-KNbSvWHQ2t4t8";

        Mono<Jwt> convertedClaims = jwtDecoder.decode(jwtWithSub);

        // THEN
        Jwt jwt = convertedClaims.block();
        assertThat(jwt.getHeaders().get("alg")).isEqualTo("HS256");
        assertThat(jwt.getSubject()).isEqualTo("123");

        /*assertThat(convertedClaims.block().getClaims())
            .containsEntry("sub", "123")
            .containsEntry("preferred_username", USERNAME)
            .containsEntry("given_name", NAME)
            .containsEntry("family_name", FAMILY_NAME)
        );*/
    }
}
```

**Pull request merged**